src/platforms/linux/daemon/apptracker.cpp

```diff
@@ -104,59 +91,116 @@ void AppTracker::userRemoved(uint userid) {
-                                const QVariantMap& extra) {
-  Q_UNUSED(display);
-  Q_UNUSED(uris);
-  Q_UNUSED(extra);
+// static
+// Expand unicode escape sequences.
+QString AppTracker::decodeUnicodeEscape(const QString& str) {
+  static const QRegularExpression re("(\\\\x[0-9A-Fa-f][0-9A-Fa-f])");
 
-  QString appIdName(appid);
-  while (appIdName.endsWith('\0')) {
-    appIdName.chop(1);
+  QString result = str;
+  qsizetype offset = 0;
+  while (offset < result.length()) {
+    QRegularExpressionMatch match = re.match(result, offset);
+    if (!match.hasMatch()) {
+      break;
+    }
+
+    bool okay;
+    qsizetype start = match.capturedStart(0);
+    QChar code = match.captured(0).mid(2).toUShort(&okay, 16);
+    if (okay && (code != 0)) {
+      result.replace(start, match.capturedLength(0), QString(code));
+      offset = start + 1;
+    } else {
+      offset = match.capturedEnd(0);
+    }
   }
-  if (!appIdName.isEmpty()) {
-    m_lastLaunchName = appIdName;
-    m_lastLaunchPid = pid;
+
+  return result;
+}
+
+// The naming convention for snaps follows one of the following formats:
+//   snap.<pkg>.<app>.service - assigned by systemd for services
+//   snap.<pkg>.<app>-<uuid>.scope - transient scope for apps
+//   snap.<pkg>.hook.<app>-<uuid>.scope - transient scope for hooks
+//
+// However, at some point the separator between the app and UUID was
+// swapped from a dot to a dash. Which makes the parsing a bit of a pain.
+//
+// See: https://github.com/snapcore/snapd/blob/master/sandbox/cgroup/scanning.go
+QString AppTracker::snapDesktopId(const QString& scope) {
+  static const QRegularExpression snapuuid(
+      "[-.][0-9a-fA-F]{8}\\b-[0-9a-fA-F]{4}\\b-[0-9a-fA-F]{4}\\b-[0-9a-fA-F]{4}"
+      "\\b-[0-9a-fA-F]{12}");
+
+  // Strip the UUID out of the scope name
+  QString stripped(scope);
+  stripped.remove(snapuuid);
+
+  // Split the remainder on dots and discard the extension.
+  QStringList split = stripped.split('.');
+  split.removeLast();
+
+  // Parse the package and application.
+  QString package = split.value(1);
+  QString app = split.value(2);
+  if (app == "hook") {
+    app = split.value(3);
   }
+  if (package.isEmpty() || app.isEmpty()) {
+    return QString();
+  }
+
+  // Reassemble the desktop identifier.
+  return QString("%1_%2.desktop").arg(package).arg(app);
 }
 
-void AppTracker::appHeuristicMatch(AppData* data) {
-  // If this cgroup contains the last-launched PID, then we have a fairly
-  // strong indication of which application this control group is running.
-  for (int pid : data->pids()) {
-    if ((pid != 0) && (pid == m_lastLaunchPid)) {
-      logger.debug() << data->cgroup << "matches app:" << m_lastLaunchName;
-      data->appId = m_lastLaunchName;
-      data->rootpid = m_lastLaunchPid;
-      break;
+// Make an attempt to resolve the desktop ID from a cgroup scope.
+QString AppTracker::findDesktopId(const QString& cgroup) {
```

Comment:
Consider adding some examples of cgroup scope names, to make it easier for the reader to understand this method. Examples for various names that start with app-gnome-, app-flatpak-, gnome-launched- etc. would be helpful.